### ISSUE TITLE
issue Latest commit wont launch any game on MacOsx 10.12 

### ISSUE THREAD

**Issue by NicolaiVdS** [citra-emu/citra#2224]
Since there is no public release for mac in a while i build my own and came to the conclusion that the latest commit wont launch any game it just stays at a black screen no errors nothing even in the terminal it does nothing.

Commit id: [3174bfd](https://github.com/citra-emu/citra/commit/3174bfd50c69dfa523671b96448113996a0bc42c)

Log from citra when launching pokemon sun/moon demo:
http://hastebin.com/vilerapaxe.log

my specs are ![Mac Specs](http://i.imgur.com/0JTX30a.png)

Issue only appears when CPU JIT is enabled in Release, in Debug CPU JIT enables works

**Comment by freiro** [comment]
I will try it on windows now to double check.
Can you please add exactly which commit are you using and your specs?

**Comment by HyperGainZ** [comment]
i use this commit `3174bfd` and my specs are ![Mac Specs](http://i.imgur.com/0JTX30a.png)

**Comment by mailwl** [comment]
error exists only on mac, release version + CPU JIT enabled. On Debug works ok

**Comment by freiro** [comment]
Pinging @Subv

**Comment by HyperGainZ** [comment]
Can confirm error only appears when in Release when CPU JIT is enabled in Debug it works fine

**Comment by HyperGainZ** [comment]
@freiro just tested a build from commit [8a73eb8](https://github.com/citra-emu/citra/commit/8a73eb802467ffeb252823ee78e4c07c91f77b37) in release and it worked il try now a build of the next commit and report back

**Comment by freiro** [comment]
The commit after that is:
https://github.com/citra-emu/citra/commit/a0b30189e5749d0bde91b09a3b64a716a2823e98
Which shouldn't involve JIT at all

**Comment by HyperGainZ** [comment]
Just did a double check and result is [8e73eb8](https://github.com/citra-emu/citra/commit/8a73eb802467ffeb252823ee78e4c07c91f77b37)  works and [593ec7f](https://github.com/citra-emu/citra/commit/593ec7fa74cf01819b0e3e29ddd3b1d3ba339090) gives just black screen when **CPU JIT** enabled 

*Note: all builds are **Release** builds **not Debug builds**

**Comment by freiro** [comment]
This is  https://github.com/citra-emu/citra/commit/593ec7fa74cf01819b0e3e29ddd3b1d3ba339090 
`-#define EMU_DATA_DIR "Citra Emulator"
+#define EMU_DATA_DIR "Citra" `

still doesn't look like the problem originates here.

**Comment by HyperGainZ** [comment]
@freiro 

> This is 593ec7f
> -#define EMU_DATA_DIR "Citra Emulator"
> +#define EMU_DATA_DIR "Citra"

aint that only apply to Windows system ?

**Comment by freiro** [comment]
it does, therefore the problem does not originate there...

**Comment by mailwl** [comment]
if i change this line back on latest master, it doesn't help

**Comment by freiro** [comment]
> if i change this line back on latest master, it doesn't help
that's to be expected I don't know why this happens in that build and not before...

**Comment by HyperGainZ** [comment]
dammit i messed up -_- [593ec7f](https://github.com/citra-emu/citra/commit/593ec7fa74cf01819b0e3e29ddd3b1d3ba339090) Works !! its [a0b3018](https://github.com/citra-emu/citra/commit/a0b30189e5749d0bde91b09a3b64a716a2823e98) that breaks stuff 
*founded out after checking the titlebar of citra -_- GG me

**Comment by JayFoxRox** [comment]
All commits in this issue so far have been pointing at changes which don't affect macOS at all. Either changes in our build script or config changes for Win32. Please bisect properly or nobody will bother to help.

**Comment by pippo2931** [comment]
It seems harmless but I believe it is #2215

**Comment by JayFoxRox** [comment]
Do you believe that or did you actually confirm that? We don't need speculation in these issues..
Despite this being a serious issue, I'm tempted to close this / remove a majority of the posts. It goes nowhere if the discussion starts out like this.

**Comment by pippo2931** [comment]
I bisected it and reverting the changes goes away.

**Comment by NicolaiVdS** [comment]
After doing my self a bisect it gave me the same result as @pippo2931  so i reverted the change and did a rebuild and tested pkmn moon to see if it launched with **CPU JIT** enabled and it worked after that i readded the test and added  **`printf("ticks: %zu / %zu \n", ticks_executed, num_instructions);`** to see what **ticks_executed** is returning and this is the result `ticks: 1586057217 / 1000` 

full console log : http://hastebin.com/ojunalasiq.log

**Comment by wwylele** [comment]
To compare with @NicolaiVdS, with the same log command, running Pokemon Sun on Linux it gives

```
ticks: 1000 / 1000 
ticks: 1002 / 1000 
ticks: 1002 / 1000 
(continue with the same number)
```

I am expecting that there is some undefined-behaviour code in Dynarmic and macOS wrongly optimized it.

**Comment by pippo2931** [comment]
When debugging in Xcode with a breakpoint on that line, I got 1000 ticks too.
It looks time sensitive.

**Comment by NicolaiVdS** [comment]
@pippo2931 if i'm not mistaken when you run citra from xcode directly its running as Debug and not as Release and there is where the problem is when building a debug version CPU JIT works but not in Release

**Comment by freiro** [comment]
For reference OOT on MSVC
https://gist.github.com/freiro/fafaa9c7166654fcc9a876d8adcd7dd6

**Comment by MerryMage** [comment]
https://github.com/MerryMage/dynarmic/blob/master/src/backend_x64/block_of_code.cpp#L39

Could someone replace:

        jit_state->cycles_remaining = cycles_to_run;
        run_code(jit_state, basic_block);
        return cycles_to_run - jit_state->cycles_remaining; // Return number of cycles actually run.

with something like:

        jit_state->cycles_remaining = cycles_to_run;
        printf("before: %zu %lld\n", cycles_to_run, jit_state->cycles_remaining);
        run_code(jit_state, basic_block);
        printf("after: %zu %lld\n", cycles_to_run, jit_state->cycles_remaining);
        return cycles_to_run - jit_state->cycles_remaining; // Return number of cycles actually run.

and share the output?

**Comment by pippo2931** [comment]
Adding the prints, it doesn't get stuck anymore.

A sample:
before: 1000 1000
after: 1000 999
before: 999 999
after: 999 995
before: 995 995
after: 995 993
before: 993 993
after: 993 0
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
after: 1000 -2
before: 1000 1000
............................
before: 185 185
after: 185 170
before: 170 170
after: 170 159
before: 159 159
after: 159 24
before: 24 24
after: 24 -1
before: 1000 1000
after: 1000 989
before: 989 989
after: 989 974
before: 974 974
after: 974 963
before: 963 963
after: 963 828
before: 828 828
after: 828 792
before: 792 792
after: 792 777
before: 777 777
after: 777 766
before: 766 766
after: 766 631

**Comment by NicolaiVdS** [comment]
same result here game just starts fine now 
output: https://nvds.be/citra_CPU_JIT_ERROR.log  (55,5MB size)

**Comment by MerryMage** [comment]
It was an ABI issue.

----

Emitted code:

**JIT prelude**

      push   %rbx
      push   %rbp
      push   %r12
      push   %r13
      push   %r14
      mov    %rdi,%r15
      stmxcsr 0x34c(%r15)
      ldmxcsr 0x348(%r15)
      jmpq   *%rsi

**JIT prologue**

      stmxcsr 0x348(%r15)
      ldmxcsr 0x34c(%r15)
      pop    %r14
      pop    %r13
      pop    %r12
      pop    %rbp
      pop    %rbx
      retq  

Register `%r15` was not being saved, despite being a callee-saved register in the ABI.

----

Here is where that code is emitted:

https://github.com/MerryMage/dynarmic/blob/57169ec0932248f703526f65095029992372a4e3/src/backend_x64/abi.cpp#L68

And `HostLocIsGPR` doesn't consider r15 a GPR:

https://github.com/MerryMage/dynarmic/blob/6520c7537e460f5e45dea837503b89814d1fb187/src/backend_x64/hostloc.h#L28-L30

**Comment by MerryMage** [comment]
Fixed in MerryMage/dynarmic@5f11b4f5.

**Comment by NicolaiVdS** [comment]
@MerryMage Just teste PR [#2243 ](https://github.com/citra-emu/citra/pull/2243)on mac and CPU JIT works again on citra & citra-qt in Release builds
